src/components/js-tracer/Canvas.js

```diff
@@ -0,0 +1,568 @@
+          );
+          nextPixelLeftSide++;
+          nextPixelRightSide++;
+          nextPixelPartialValue = 0;
+        }
+      };
+
+      for (let i = 0; i < timing.length; i++) {
+        const eventStartTime = timing.start[i];
+        const eventEndTime = timing.end[i];
+
+        // Only draw samples that are in bounds.
+        if (
+          eventEndTime > timeAtViewportLeft &&
+          eventStartTime < timeAtViewportRightPlusMargin
+        ) {
+          const unitIntervalStartTime: UnitIntervalOfProfileRange =
+            (eventStartTime - rangeStart) / rangeLength;
+          const unitIntervalEndTime: UnitIntervalOfProfileRange =
+            (timing.end[i] - rangeStart) / rangeLength;
+
+          let x: DevicePixels =
+            (unitIntervalStartTime - viewportLeft) *
+              devicePixels.innerContainerWidth /
+              viewportLength +
+            devicePixels.timelineMarginLeft;
+          const uncutWidth: DevicePixels =
+            (unitIntervalEndTime - unitIntervalStartTime) *
+            devicePixels.innerContainerWidth /
+            viewportLength;
+          const eventIndex = timing.index[i];
+          const isHovered = hoveredItem === eventIndex;
+          if (uncutWidth === 0) {
+            continue;
+          }
+
+          let w = uncutWidth;
+          if (x + w < devicePixels.timelineMarginLeft) {
+            continue;
+          }
+          if (
+            x >
+            devicePixels.containerWidth - devicePixels.timelineMarginRight
+          ) {
+            continue;
+          }
+          if (x < devicePixels.timelineMarginLeft) {
+            // Adjust events that are before the left margin.
+            w = w + x - devicePixels.timelineMarginLeft;
+            x = devicePixels.timelineMarginLeft;
+          }
+
+          const text = timing.label[i];
+          if (isHovered) {
+            hoveredElement = { x, y, w, h, uncutWidth, text };
+            continue;
+          }
+
+          const ceilX = Math.ceil(x);
+          if (ceilX < x + w) {
+            if (x >= nextPixelRightSide) {
+              // This value skips the partial pixel. Commit that last partial pixel
+              // |0---1|1---2|2---3|3---4|4---5|
+              //  XXXXX XXXXX  0.2
+              //                             [4.9---6.2]
+              commitPartialPixel(y);
+              nextPixelLeftSide = ceilX - 1;
+              nextPixelRightSide = ceilX;
+            } else {
+              // |0---1|1---2|2---3|3---4|4---5|
+              //  XXXXX XXXXX  0.2
+              //                 [2.9------6.2]
+              //
+              // or
+              //
+              // |0---1|1---2|2---3|3---4|4---5|
+              //  XXXXX XXXXX  0.2
+              //                   [3------6.2]
+            }
+            const leftOfPixelDivide = ceilX - x;
+            x = ceilX;
+            nextPixelPartialValue += leftOfPixelDivide;
+            commitPartialPixel(y);
```

Comment:
I don't understand why we commit the partial pixel here. Maybe this will get clearer once you reorganize the code as advised in my other comments.